Retrieve RISC-V Version String job

M-Davies opened the pull request:
* Instead of retrieving the faked version string, we run a downstream job on a riscv machine after the build that runs java version and passes it back to the pipeline in time for parseVersionOutput()
* Also clears up white spaces in openjdk_build_pipeline.groovy

Fixes: #1773 
Signed-off-by: Morgan Davies <morgandavies2020@gmail.com>

sxa reviewed (commented):
Just because I'm picky ... Can you change the title of the PR and the commit message to write the architecture as `RISC-V` please?

M-Davies commented:
> Just because I'm picky ... Can you change the title of the PR and the commit message to write the architecture as `RISC-V` please?

Sure but it'll still appear as `riscv` in the code as that's what the architecture is listed as in the config file

karianna commented:
Can this come out of draft now?

M-Davies commented:
Hitting a node issue at the moment. Java version will not run in the azure docker container, for ref see https://ci.adoptopenjdk.net/job/build-scripts/job/utils/job/riscv-version-out (builds 12-14):
```
13:45:16  ./java: 1: ./java: ELF��: not found
13:45:16  ./java: 2: ./java: @h@8: not found
13:45:16  ./java: 3: ./java: : not found
13:45:16  ./java: 3: ./java: 0: not found
13:45:16  ./java: 20: ./java: Syntax error: Bad function name
```
I'm pretty sure it's running on the binary due to all the debug I'm spitting out to verify it but I get this error no matter where I execute the binary from. It's likely the environment (since it's actually running on the azure build agent, not the riscv machine itself) which begs the question, which riscv machine am I supposed to use from [this list](https://ci.adoptopenjdk.net/label/riscv/)?

sxa commented:
I would literally go with "anything that's online" which at the moment is just one from that list - so just have it select the `riscv` label and it'll pick whatever we happen to have online at the time. We don't have enough RISC-V systems to get picky about it ;-)

M-Davies commented:
> I would literally go with "anything that's online" which at the moment is just one from that list - so just have it select the `riscv` label and it'll pick whatever we happen to have online at the time. We don't have enough RISC-V systems to get picky about it ;-)

I am already picking the `riscv` label and it picks up the docker container every time which is why it's failing. I'll try and pick a node from that list and go with that then

sxa commented:
> > I would literally go with "anything that's online" which at the moment is just one from that list - so just have it select the `riscv` label and it'll pick whatever we happen to have online at the time. We don't have enough RISC-V systems to get picky about it ;-)
> 
> I am already picking the `riscv` label and it picks up the docker container every time which is why it's failing. I'll try and pick a node from that list and go with that then

Hmmm OK we need to fix that - the x64 machines used to build should NOT have `riscv` on them - I set up a tag of `riscvcross` when we were buiding natively.

But if that's been broken use `riscv&&ci.role.test`

M-Davies commented:
Note when using the jenkins `sh` tag and `java -version`, make sure you specify `2>&1` to redirect to stdout, even if `returnStdout` is set 🤦

M-Davies commented:
run tests

M-Davies commented:
> # 🟠 PR TESTER RESULT 🟠
> ❎ Some pipelines failed or the job was aborted! ❎
> _See the `pipeline-build-check` below for more information..._

Failed due to existing OpenJ9 problems https://github.com/eclipse/openj9/issues/11236

M-Davies commented:
> 1) JSON files aren't being produced (due to early archive)
> * So we try archiving the binaries first, and then the metadata

Fixed by archiving the binaries, and then the metadata 
> 2) JSON data shows it's using the faked version.txt file and not new one
> * Change name of faked version to ensure it's not picked up
> * Add additional check in formMetadata() to look for riscv info

Fixed by description above

See https://ci.adoptopenjdk.net/view/work-in-progress/job/build-scripts-testing/job/jobs/job/jdk11u/job/jdk11u-linux-riscv64-openj9/26/ for evidence this works. I'll run tests one more time to ensure that I haven't broken anything new and open this up for review

sxa reviewed (commented):
Since it's Friday evening I'll try and do a review of this next week (We won't want to merge it tonight regardless!) but thanks for working on this and great to see it seemingly working as it is quite possible that we'll need this for other platforms in the future :-D

M-Davies commented:
> # 🟠 PR TESTER RESULT 🟠
> ❎ Some pipelines failed or the job was aborted! ❎
> _See the `pipeline-build-check` below for more information..._

Failed due to issues with connecting to the docker agent. Existing issue unrelated to this PR https://github.com/AdoptOpenJDK/openjdk-build/issues/2244
```
17:35:46  Compiling /home/adoptopenjdk/workspace/build-scripts-pr-tester/build-test/jobs/jdk8u/jdk8u-linux-x64-hotspot/workspace/build/src/hotspot/src/share/vm/code/dependencies.cpp
17:35:47  Cannot contact azurebuildagentd92090: hudson.remoting.RequestAbortedException: java.io.IOException: Unexpected termination of the channel
01:28:39  Cancelling nested steps due to timeout
01:28:39  Could not connect to azurebuildagentd92090 to send interrupt signal to process
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
01:28:39  [ERROR] Build JDK timeout (8 HOURS) has been reached. Exiting...
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
01:28:39  ERROR: Issue with creating launcher for agent azurebuildagentd92090. The agent has not been fully initialized yet
```

karianna commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -583,9 +584,10 @@ class Build {
-                context.println "ERROR: ${versionPath} was not found. Exiting..."
-                throw new Exception()
+            } catch (Exception e) {
+                context.println "WARNING: ${versionPath} was not found. Likelihood is this is a riscv build which creates a different version file. Attempting to read ${riscvVersionPath}"
```
This is a little bit of a red herring - can we detect if we are meant to be looking for riscv and predicate some of this debug accordingly?

M-Davies replied:
I figured asking for forgiveness was better than asking for permission. Nevertheless, it should be pretty easy to adapt to look for riscv

karianna commented on pipelines/build/common/riscv_version_out.groovy:
```diff
@@ -0,0 +1,116 @@
+node ("riscv&&ci.role.test") {
+    timestamps {
+        try {
+            Integer JOB_TIMEOUT = 1
```
I suspect some of this can be extracted to local functions/methods or does the DSL prevent this?

M-Davies replied:
I stuck it in a separate file mostly to avoid clogging up `openjdk_build_pipeline.groovy`, as that one is quite big already. It doesn't have to be and can be moved across if people think it's better keeping it all together?

M-Davies commented:
run tests

M-Davies commented:
Running one final test at https://ci.adoptopenjdk.net/view/work-in-progress/job/build-scripts-testing/job/jobs/job/jdk11u/job/jdk11u-linux-riscv64-openj9/27/console but assuming that passes and no more changes are requested, this should be good to merge at an admin discretions

gdams commented:
One observation (and it doesn't need to be done in this PR). Are you able to generalise this job to be for any cross platform builds. I'm thinking of windows arm64 and macOS arm64 which will likely need the same logic

sxa commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -578,11 +579,14 @@ class Build {
 
             // Get Full Version Output
             String versionPath = "workspace/target/metadata/version.txt"
+            if (buildConfig.ARCHITECTURE.contains("riscv")) {
```
I wonder if we can modify this conditional to be more of a generic "If we are cross-compiling" (true for any RIS-V specific references in method names etc. here) but that's 'somethign we can change later as we don't have an obvious way to detect it at the moment so would be quite a bit more work :-)

M-Davies replied:
I think that (and @gdams [suggestion above](https://github.com/AdoptOpenJDK/openjdk-build/pull/2238#issuecomment-733633743)) should probably be done under a separate PR once we have this job running for linux and we're happy with it

sxa commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -819,6 +823,48 @@ class Build {
     }
 
+    /*
+    Run the Riscv version reader downstream job.
```
```suggestion
    Run the RISC-V version reader downstream job.
```

sxa commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -819,6 +823,48 @@ class Build {
+        // Archive the artifacts early so we can copy them over to the downstream job
+        try {
+            context.timeout(time: buildTimeouts.BUILD_ARCHIVE_TIMEOUT, unit: "HOURS") {
+                context.archiveArtifacts artifacts: "workspace/target/*"
```
I'm assuming this doesn't stop the normal archive at the end  which will still be needed to pick up the metadata

M-Davies replied:
Yes that happens regardless

sxa commented on pipelines/build/common/riscv_version_out.groovy:
```diff
@@ -0,0 +1,116 @@
+
+// TODO: ADD THE ACTIVE NODE TIMEOUT LOGIC HERE OR GET IT MERGED INTO JOB HELPER (https://github.com/AdoptOpenJDK/openjdk-build/issues/2235)
+
+node ("riscv&&ci.role.test") {
```
The build machines used for cross compiling shouldn't have the `riscv` tag on them (but as discused elsewhere they seem to at the moment) but we can roll with this for now and update it later

M-Davies replied:
That's probably worth opening an issue for, as it's worrying they have been labelled as such when they actually aren't

sxa replied:
@M-Davies I can see machines with `riscv64` but not `riscv` - did you hit some problem using `riscv`? I've removed `riscv64` from one of the static machines (which is offline anyway) but I'm surprised that it caused a problem with the current machine sets and labels

M-Davies replied:
I don't think I tried `riscv64` as I didn't realise it existed. The issue I had with `riscv` on it's own was that it always `node` into an azure build agent instead of an actual riscv machine. Adding `ci.role.test` ensured it picked up one of George's machines

sxa commented on pipelines/build/common/riscv_version_out.groovy:
```diff
@@ -0,0 +1,116 @@
+                    sh "tar -zxvf ${jdkFileFilter} && rm ${jdkFileFilter}"
+
+                    String jdkDir = sh(
+                        script: "ls | grep jdk*",
```
`*` is superfluous here - suggest removing it to avoid potential confusion (Could also use `ls -d *jdk*`
```suggestion
                        script: "ls | grep jdk",
```

sxa commented on sbin/build.sh:
```diff
@@ -538,11 +538,12 @@ printJavaVersionString() {
        local jdkversionNoPrefix=${jdkversion#jdk-}
        local jdkShortVersion=${jdkversionNoPrefix%%+*}
-       cat << EOT > "${BUILD_CONFIG[WORKSPACE_DIR]}/${BUILD_CONFIG[TARGET_DIR]}/metadata/version.txt"
+       cat << EOT > "${BUILD_CONFIG[WORKSPACE_DIR]}/${BUILD_CONFIG[TARGET_DIR]}/metadata/fakedVersion.txt"
```
I guess this can be removed now since it should never be required? Again, not critical and we can leave it in for now if it won't cause problems but perhaps a comment indicating it is no longer required as a minimum

M-Davies replied:
Done

sxa reviewed (approved):
Approved. Would be good to merge this today and kick off a pipeline just to validate it in production. Comments should be considered suggestions only so I've approved regardless - if you want to include any changes based on my comments feel free but I won't be hurt if you don't :-)

M-Davies commented:
run tests
